feat: add genesis download functionality for known chains

deblasis opened the pull request:
## Overview

This PR aims at addressing #4891 

Specifically:
- Add genesis download with SHA-256 verification for known chains (https://github.com/celestiaorg/celestia-app/issues/4891#issuecomment-2921590680) 
- New download-genesis CLI command
- Auto-download during init for public chains with warnings
- Support celestia, mocha-4, arabica-10, arabica-11 networks


Fixes #4891

![image](https://github.com/user-attachments/assets/4b4c73e8-59bc-4e0e-9575-665d2c2a295c)

rach-id commented on app/download_genesis_test.go:
```diff
@@ -0,0 +1,395 @@
+
+				// Verify file content
+				content, err := os.ReadFile(tt.filepath)
+				if err != nil {
+					t.Errorf("Failed to read downloaded file: %v", err)
+					return
+				}
```
same

rach-id commented on app/download_genesis_test.go:
```diff
@@ -0,0 +1,395 @@
+					return
+				}
+
+				if string(content) != tt.expectContent {
+					t.Errorf("downloadFile() content = %s, want %s", string(content), tt.expectContent)
+				}
```
same

rach-id commented on app/download_genesis.go:
```diff
@@ -0,0 +1,119 @@
+
+	url := fmt.Sprintf("%s/%s/genesis.json", genesisBaseURL, chainID)
+	if err := downloadFile(outputFile, url); err != nil {
+		return fmt.Errorf("error downloading / persisting the genesis file: %s", err)
```
```suggestion
		return fmt.Errorf("error downloading / persisting the genesis file: %w", err)
```

deblasis commented:
> thanks for addressing the feedback so quickly. left some code improvements

Sorry for the silly mistakes. I wrapped the errors and used testify to simplify the tests. 

Also I fixed a test in the **cmd** package that I had missed (I caused the failure with my refactoring, now it's OK).

deblasis commented on app/download_genesis.go:
I felt like the logic was quite complex and domain specific and normally I tend to keep my `cmd` packages focused on managing the CLI, outsourcing complexity in the relevant packages.

As soon as I saw the chains and the SHAs of the genesis files in the cmd package I wanted to relocate that code elsewhere if that makes sense.

deblasis commented on app/download_genesis_test.go:
```diff
@@ -0,0 +1,364 @@
+	for _, chainID := range expectedChains {
+		if !strings.Contains(result, chainID) {
+			t.Errorf("ChainIDs() = %s, should contain %s", result, chainID)
+		}
```
https://github.com/celestiaorg/celestia-app/pull/5024#discussion_r2164882595

rach-id replied:
same https://github.com/celestiaorg/celestia-app/pull/5024#discussion_r2165927543

deblasis commented on app/download_genesis_test.go:
```diff
@@ -0,0 +1,364 @@
+			result := GetChainIDOrDefault(tt.args)
+			if result != tt.expected {
+				t.Errorf("GetChainIDOrDefault(%v) = %s, want %s", tt.args, result, tt.expected)
+			}
```
https://github.com/celestiaorg/celestia-app/pull/5024#discussion_r2164882595

deblasis commented on app/download_genesis_test.go:
```diff
@@ -0,0 +1,364 @@
+				content, err := os.ReadFile(tt.filepath)
+				require.NoError(t, err)
+
+				require.Equal(t, tt.expectContent, string(content))
```
https://github.com/celestiaorg/celestia-app/pull/5024#discussion_r2164882595

deblasis commented on app/download_genesis_test.go:
```diff
@@ -0,0 +1,364 @@
+			err := downloadFile(tt.filepath, tt.url)
+
+			if tt.wantErr {
+				require.Error(t, err)
```
https://github.com/celestiaorg/celestia-app/pull/5024#discussion_r2164882595

rach-id commented on app/download_genesis_test.go:
```diff
@@ -0,0 +1,364 @@
+			result := IsKnownChainID(tt.chainID)
+			if result != tt.expected {
+				t.Errorf("IsKnownChainID(%s) = %v, want %v", tt.chainID, result, tt.expected)
+			}
```
but for the comment here, is was for refactoring:
```
			if result != tt.expected {
				t.Errorf("IsKnownChainID(%s) = %v, want %v", tt.chainID, result, tt.expected)
			}
```

to:
```
assert.Equal(t, tt.expected, result)
```

deblasis replied:
I see, thank you for taking the time to elaborate more 👍

rach-id commented on app/download_genesis_test.go:
```diff
@@ -0,0 +1,364 @@
+	// Check that result is comma-separated
+	if !strings.Contains(result, ",") {
+		t.Errorf("ChainIDs() = %s, should be comma-separated", result)
+	}
```
https://github.com/celestiaorg/celestia-app/pull/5024#discussion_r2165927543 same for the rest

rach-id commented on app/download_genesis.go:
```diff
@@ -0,0 +1,119 @@
+// To add a new chain-id, download the genesis file from the networks repo and compute the SHA-256 hash.
+// Add the chain-id and hash to this map.
+var chainIDToSha256 = map[string]string{
+	"celestia":   "9727aac9bbfb021ce7fc695a92f901986421283a891b89e0af97bc9fad187793",
+	"mocha-4":    "0846b99099271b240b638a94e17a6301423b5e4047f6558df543d6e91db7e575",
+	"arabica-10": "fad0a187669f7a2c11bb07f9dc27140d66d2448b7193e186312713856f28e3e1",
+	"arabica-11": "77605cee57ce545b1be22402110d4baacac837bdc7fc3f5c74020abf9a08810f",
```
where did you get these hashes btw? 

mocha-4 doesn't seem to be the correct hash: https://github.com/celestiaorg/networks/blob/master/mocha-4/genesis_hash.txt

deblasis replied:
I took them from here:

https://github.com/celestiaorg/celestia-app/blob/e4f57db0a67a326ec9cdced4ccce6a5a595252b3/cmd/celestia-appd/cmd/download_genesis.go#L19-L24

rach-id reviewed (dismissed):
utAck

deblasis commented:
Sorry to bother you again @rach-id 🙈
 I noticed that I forgot to lint and also there was a typo I introduced while moving the code that made a test fail. It should be all sorted now.

deblasis commented:
@rach-id: If I understood correctly how Tastora works, this PR https://github.com/celestiaorg/tastora/pull/55 should be enough to handle the new flag and allow E2E tests to pass with a nice green

deblasis commented:
> also, since the change in the other PR is just a single line, and it would make the CI green for this PR. Please move that change to this PR.

The PR https://github.com/celestiaorg/tastora/pull/55 is on another repository, I can't bring it here 😄 
Unless you know another way to pass in the flag to the init command in `tastora` from here I didn't notice.

deblasis commented:
@rach-id I guess we have a Chicken and Egg problem. Which one came first? 🤣 

Essentially: 

- in this PR we need https://github.com/celestiaorg/tastora/pull/55 to be merged to have successful E2E tests
- in https://github.com/celestiaorg/tastora/pull/55 `tastora` needs the `celestia-app` binary to support the new `--yes` flag

I mean, these are my first contributions and I am already having fun!

I guess we need to coordinate.

I suggest we proceed with PR review and to make it essentially ready to be merged, same with the PR in the other repository, then when we are ready we should merge with a commit message that skips the CI run on both repositories: `[skip ci]`.

Once both PRs are merged we should run CI on both and expect all green, otherwise we should fix/revert the merges and try again.

Of course this is just the way I would do it, but I might be totally wrong.

deblasis commented:
I looked into the way you guys test things, pretty cool stuff. I realized that the environment variable is not passing thought because in k8s it's a completely separate environment so in 4d30823cad47347e4db0c1e459687d208ecb809b I set it myself.

Unless I missed something else, this should be finally it.

rach-id commented:
I'm beginning to lean towards not having an interactive yes. And just download by default the genesis for known chains. However, I would like @rootulp opinion on this.

deblasis commented:
Hi guys.

The latest change worked! Using the environment variable to toggle non-interactivity worked.

This means that there's no need for the other PR https://github.com/celestiaorg/tastora/pull/55 (I just closed it)

## Receipts

### Before
![image](https://github.com/user-attachments/assets/20d81ba7-2aff-41e4-b832-c99ce27d9524)
> The tests were failing waiting for the user to interact with it

### After
![image](https://github.com/user-attachments/assets/ae67de77-f037-4c38-b7e6-207ffaf4b7d6)
> the `Init` downloads the genesis 🤘 (we don't see logs because the test doesn't capture stdout inside the github action, perhaps one improvement here could be to add log entries when running in CI so that we know what is happening when not interactive)

The tests failed for other reasons, I have the same failure locally if I replicate the steps that the CI performs, so this is totally independent from the interactivity toggle:
![image](https://github.com/user-attachments/assets/34f66c7e-da7b-4733-9373-8a59b704f6b0)


I am obviously going to get rid of interactivity if you are all on the same page, but I am VERY curious about the **why** you backtracked on what sounded like a legit requirement. 👀

deblasis commented:
> @deblasis Thanks a lot for sticking around, get involved in discussions, and apply review feedback so quickly 🙏

You are welcome @rach-id 

This was fun!

rach-id commented on cmd/celestia-appd/cmd/init.go:
```diff
@@ -31,6 +31,8 @@ const (
 	FlagOverwrite = "overwrite"
 	// FlagRecover defines a flag to initialize the private validator key from a specific seed.
 	FlagRecover = "recover"
+	// FlagConfirmKnownChain defines a flag that automatically answers yes to prompts for known/public chainIDs
```
[nit]
```suggestion
	// FlagConfirmKnownChain defines a flag that automatically answers yes to prompts for known/public chainIDs.
```

rach-id commented on test/docker-e2e/e2e_test.go:
```diff
@@ -65,7 +66,7 @@ func (s *CelestiaTestSuite) CreateDockerProvider(opts ...ConfigOption) celestiat
 			NumValidators: &numValidators,
 			NumFullNodes:  &numFullNodes,
-			ChainID:       "celestia",
+			ChainID:       appconsts.TestChainID,
```
using the chainID "celestia" confuses whether this test is connecting to mainnet or a local network. Also, if you want to init a genesis with a known chain, the confirm known chain needs to be true.

rootulp commented on cmd/celestia-appd/cmd/init.go:
```diff
@@ -104,6 +106,41 @@ func initCmd(basicManager module.BasicManager, defaultNodeHome string) *cobra.Co
+				if !confirm {
+					inBuf := bufio.NewReader(cmd.InOrStdin())
+					var err error
+					confirm, err = input.GetConfirmation("Do you want to download the known genesis for this chain?", inBuf, cmd.OutOrStdout())
```
Confirmation prompts are a UX inconvenience because a node operator can't copy + paste a series of commands into their terminal. IMO if a confirmation prompt is necessary than it would be preferable for the binary to not auto download anything for known chain-ids and expect the node operator to do

```
celestia-appd init celestia
celestia-appd download-genesis
```

instead of

```
celestia-appd init celestia
<wait>
y
```

rootulp commented on cmd/celestia-appd/cmd/download_genesis_test.go:
```diff
@@ -1,27 +1,357 @@
+func TestDownloadGenesis(t *testing.T) {
+	tempDir := t.TempDir()
+
+	// Content that matches the expected hash for celestia
+	correctContent := `{"genesis_time":"2023-01-01T00:00:00Z","chain_id":"celestia"}`
+	correctHasher := sha256.New()
+	correctHasher.Write([]byte(correctContent))
+	correctHash := hex.EncodeToString(correctHasher.Sum(nil))
```
I think it's unnecessary to test made up hashes. I don't think these test cases add much value.

rootulp commented on cmd/celestia-appd/cmd/init.go:
```diff
@@ -104,6 +106,41 @@ func initCmd(basicManager module.BasicManager, defaultNodeHome string) *cobra.Co
+					}
+				}
+
+				if !confirm {
+					fmt.Println("Genesis download cancelled. For local development, please use a different chain-id (e.g., 'test-chain-1').")
+					return nil
+				}
```
Can this be merged into the conditional block above?

deblasis commented on cmd/celestia-appd/cmd/download_genesis_test.go:
```diff
@@ -1,27 +1,357 @@
+			name:     "empty chain ID",
+			chainID:  "",
+			expected: false,
+		},
```
I appreciate the feedback, but I respectfully disagree. Let me try again discussing the testing approach here and my philosophy in general. 

The current test already covers the core functionality adequately by testing:

- Known chain ID → true
- Unknown chain ID → false
- Empty chain ID → false

Adding all specific chain IDs would create several issues:

1. **Coupling**: The test becomes tightly coupled to the current data set. When new chains are added (which seems frequent given arabica-10 vs arabica-11), we'd need to update tests purely for data changes, not logic changes.
2. **Testing data vs. behavior**: `isKnownChainID` is a simple map lookup function. We should test that the lookup logic works correctly, not validate the completeness of our data set.
3. **Maintenance overhead**: Every chain addition/removal becomes a test change, creating noise in the test suite without testing additional functionality.

The mock approach I suggested would:

- Test the actual function behavior **deterministically**
- Remain stable regardless of chain ID changes, making it essentially future-proof
- Focus on the algorithm rather than current data
- Follow testing best practices of isolating units under test

If we want to ensure all current chains are covered that's another story, we could add a separate test that validates the completeness of `chainIDToSha256`, keeping our unit test focused on behavior.

What are your thoughts on this distinction between testing behavior vs. testing data completeness?

rootulp replied:
> The test becomes tightly coupled to the current data set. When new chains are added (which seems frequent given arabica-10 vs arabica-11), we'd need to update tests purely for data changes, not logic changes.

We don't add chain IDs frequently so IMO it would be more beneficial to have exhaustive test coverage of all the known chain IDs. I think it's good idea to add a test case every time we add a chain ID to the map.

rootulp commented:
Overall, thank you for persisting through several rounds of feedback. It is my fault the original issue underspecified some details that only became apparent after iterations on this PR. Given our team doesn't have consensus on details for the feature (e.g. should the automatic download require a confirmation prompt?), it is unfair to expect you to drive this forward.

Given it isn't a productive use of time to debate the rationale for some of the unit tests in this PR, I'm inclined to close for now.

deblasis commented:
The beauty of open source lies in finding consensus (blockchain pun intended) and iterating through feedback. I was happy to proceed with the team's decision, whether or not it included the interactivity flag. 

Asking questions is in my nature, otherwise, I’d be an LLM that just writes code and executes commands until the PR is merged. 🤣 

I think you are wrong. Debating the rationale behind unit tests is a productive use of time, especially when there’s disagreement and you should expect pushback. It’s how we learn and improve. It's a mistake to assume we know everything because we do not.

Thank you all, it’s been a valuable experience, and I enjoyed contributing. 🫡
